### PR TITLE
feat: `HeaderBase` migration (extension)

### PR DESCRIPTION
**Pull request by kirillzyusko**
## **Description**

Added `Header` component to DSR.

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/DSYS-303

## **Manual testing steps**

1. Open Storybook
2. Check `HeaderBase` component

### **Before**

image

### **After**

Screenshot 2026-05-01 at 13 41 08

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it introduces a new exported layout primitive and changes centering behavior to CSS Grid; consumers relying on legacy Box-polymorphism/class hooks or edge-case sizing may see layout differences.
> 
> **Overview**
> Adds a new `HeaderBase` component to `@metamask/design-system-react`, exported via the components barrel, with Storybook docs/stories and a test suite.
> 
> `HeaderBase` implements centered-title + start/end accessory layout using a **3-column CSS Grid** (no JS measurement/resize listeners) and provides wrapper prop bags for each slot with `className` merging via `twMerge`.
> 
> Updates `MIGRATION.md` with extension-to-design-system guidance for `HeaderBase`, including breaking changes (no polymorphic root/Box utility props, removal of `mm-header-base` hook) and before/after examples.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 870dd58e3fcb3786ea396c678b034794e0aafd7e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/design-system-react/src/components/HeaderBase/HeaderBase.types.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import type { BoxProps } from '../Box';
+
+/**
+ * Box prop bag for the three internal slot wrappers (start accessory, title,
+ * end accessory). Mirrors the legacy `BoxProps<'div'>` typing but is bound to
+ * the MMDS `BoxProps` shape and carries an explicit `data-*` index signature
+ * so consumers can pass test ids and other dataset attributes through
+ * `Partial`/`Omit` indirection.
+ */
+type SlotWrapperProps = Omit<BoxProps, 'children'> & {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/HeaderBase/HeaderBase.types.ts]: suggestion: Can we exclude `asChild` from `SlotWrapperProps` here? `Omit<BoxProps, 'children'>` still leaks the Box `asChild` escape hatch into `childrenWrapperProps`, `startAccessoryWrapperProps`, and `endAccessoryWrapperProps`, but these wrappers only exist to host arbitrary `ReactNode` slot content. If a consumer passes `asChild: true`, Box switches to Radix `Slot`, which requires a single element child, so valid `HeaderBase` content like strings, fragments, or multiple nodes can typecheck and then fail at runtime. `Omit<BoxProps, 'children' | 'asChild'>` seems closer to the intended contract.

File: packages/design-system-react/src/components/HeaderBase/HeaderBase.types.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  [key: `data-${string}`]: string | undefined;
+};
+
+export type HeaderBaseProps = Omit<ComponentProps<'div'>, 'children'> & {
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/components/HeaderBase/HeaderBase.types.ts:16]: non-blocking: Compared to `packages/design-system-react-native/src/components/HeaderBase/HeaderBase.types.ts`, this is still a different `HeaderBase` contract rather than a platform-specific layer over a shared one. Per `.cursor/rules/component-architecture.md`, I think this family should eventually converge on a shared `HeaderBasePropsShared` source of truth in `design-system-shared`, with platform-only differences layered on top. I don’t think this PR necessarily needs to absorb that full alignment to unblock, but it seems worth calling out before more header-family components build on top of both surfaces.

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
I compared this migration against the extension `header-base` implementation, the existing React Native `HeaderBase`, and the component-architecture rules. The grid-based rewrite and migration docs look good overall, and the focused `HeaderBase` test file passes on the current head. I found one current runtime-contract issue in the slot wrapper prop bags, plus one non-blocking cross-platform architecture follow-up.

**Review by georgewrmarshall — APPROVED** [review]
LGTM 🚀
